app/models/session.rb

```diff
@@ -11,21 +11,39 @@ def self.check_session_timeout
     purge(ttl)
   end
 
-  def self.purge(ttl)
-    ses = where("updated_at <= ?", ttl.seconds.ago.utc)
-    ses.each do|s|
-      begin
-        userid = Marshal.load(Base64.decode64(s.data.split("\n").join))[:userid]
-        user = User.find_by_userid(userid)
-        if user && ((user.lastlogoff && user.lastlogon && user.lastlogoff < user.lastlogon) || (user.lastlogon && user.lastlogoff.nil?))
-          user.logoff
-        end
-      rescue Exception
-      end
+  def self.purge(ttl, batch_size = 100)
+    deleted = 0
+    loop do
+      cnt = purge_one_batch(ttl, batch_size)
+      deleted += cnt
+
+      break if cnt.zero?
+    end
+
+    _log.info("purged stale session data, #{deleted} entries deleted") unless deleted == 0
+  end
+
+  def self.purge_one_batch(ttl, batch_size)
+    sessions = where("updated_at <= ?", ttl.seconds.ago.utc).limit(batch_size)
+    return 0 if sessions.size.zero?
 
-      s.destroy
+    log_off_user_sessions(sessions)
+    where(:id => sessions.collect(&:id)).destroy_all.size
+  end
+
+  def self.log_off_user_sessions(sessions)
+    # Log off the users associated with the sessions that are eligible for deletion
+    userids = sessions.each_with_object([]) do |s, a|
+      a << Marshal.load(Base64.decode64(s.data.split("\n").join))[:userid]
+    end
+
+    User.where(:userid => userids).each do |user|
+      if ((user.lastlogoff && user.lastlogon && user.lastlogoff < user.lastlogon) || (user.lastlogon && user.lastlogoff.nil?))
+        user.logoff
+      end
     end
-    _log.info("purged stale session data, #{ses.length} entries deleted") unless ses.length == 0
+  rescue => err
+    _log.warn("Error '#{err.message}', attempting to delete session with id [#{sessions.id}]")
```

Comment:
@gtanzillo the error message looks wrong.  Isn't sessions an array of sessions?  It won't have an id, right?

What are we catching here?  The Marshal.load line or the user.logoff?